lib/puppet/settings.rb

```diff
@@ -52,13 +52,14 @@ class Puppet::Settings
   # returns reasonable application default settings values for a given run_mode.
   def self.app_defaults_for_run_mode(run_mode)
     {
-        :name     => run_mode.to_s,
-        :run_mode => run_mode.name,
-        :confdir  => run_mode.conf_dir,
-        :codedir  => run_mode.code_dir,
-        :vardir   => run_mode.var_dir,
-        :rundir   => run_mode.run_dir,
-        :logdir   => run_mode.log_dir,
+        :name      => run_mode.to_s,
+        :run_mode  => run_mode.name,
+        :confdir   => run_mode.conf_dir,
+        :codedir   => run_mode.code_dir,
+        :vardir    => run_mode.var_dir,
+        :publicdir => run_mode.public_dir,
```

Comment:
These defaults are only for "root" directories. If you default `publicdir` to `"$vardir/public"` instead of nil, then you won't need to add `publicdir` here.

Reply:
That was our initial intention but the problem is that `$vardir` is actually `/opt/puppetlabs/puppet/cache` which seems to have `0710`\`drwxr-x---` permissions (as pointed out also by @gimmyxd in [this](https://tickets.puppetlabs.com/browse/PA-3253?focusedCommentId=744843&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-744843) comment) so public class users wouldn't get past this folder if we added any `755`\`rwxr-xr-x` folders\files in it. The obvious solution would be changing the `vardir`'s and its subfolders permissions according to our given needs but we wouldn't want to arise any security concerns (missing folders, different permissions behavior on Windows or just clients relying on the folder's permission for storing their own custom files\folders there) so we decided that making the `public` folder be on the same level as the `cache` folder would be the best\safest approach. Would like to hear further thoughts/ideas/concerns on this topic 😃.